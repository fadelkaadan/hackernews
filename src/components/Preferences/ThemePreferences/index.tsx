import React from "react";
import styled, { keyframes } from "styled-components";
import { useDispatch, useSelector } from "react-redux";
import { RootState } from "../../../store/rootReducer";
import { setTheme } from "../../../store/preferences/actions";
import { ThemeEnum } from "../../../store/preferences/types";

const Wrapper = styled.div`
  grid-column: content;
  display: flex;
  align-items: center;
`;

const animation = keyframes`
  0% {
		background-position: 0% 50%;
	}
	50% {
		background-position: 100% 50%;
	}
	100% {
		background-position: 0% 50%;
	}
`;

const Circle = styled.div`
  height: 30px;
  width: 30px;
  border-radius: 50%;
  border: none;
  outline: none;
  margin-left: 20px;
  cursor: pointer;
  box-sizing: border-box;

  transition: 0.1s;
  background-size: 400% 400%;
  animation: ${animation} 5s ease infinite;
  opacity: 0.7;

  &:hover {
    opacity: 1;
  }
`;

const Title = styled.h3``;

const Dark = styled(Circle)`
  background-image: linear-gradient(
    to right bottom,
    #282c2e 0%,
    #404449 50%,
    #2d3031 100%
  );
`;

const Night = styled(Circle)`
  background-image: linear-gradient(
    to right bottom,
    #355397 0%,
    #233964 51%,
    #4b6cb7 100%
  );
`;

const Light = styled(Circle)`
  background-image: linear-gradient(
    to right,
    #f3f3f3 0%,
    #c0c0c0 50%,
    #e9e9e9 100%
  );
`;

const ThemePreferences = () => {
  const dispatch = useDispatch();
  const show = useSelector(
    (state: RootState) => state.preferences.isPreferencesClicked
  );
  const changeTheme = (theme: ThemeEnum): void => {
    dispatch(setTheme(theme));
  };
  return show ? (
    <Wrapper>
      <Title>Theme</Title>
      <Dark onClick={() => changeTheme(ThemeEnum.DARK)} />
      <Night onClick={() => changeTheme(ThemeEnum.NIGHT)} />
      <Light onClick={() => changeTheme(ThemeEnum.LIGHT)} />
    </Wrapper>
  ) : null;
};

export default ThemePreferences;
