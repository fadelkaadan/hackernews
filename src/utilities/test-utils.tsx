import React from "react";
import { Provider } from "react-redux";
import ThemeProvider from "../theme/ThemeProvider";
import { BrowserRouter as Router } from "react-router-dom";
import configureStore from "redux-mock-store";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import "@testing-library/jest-dom/extend-expect";
import { ThemeEnum } from "../store/preferences/types";
import { RootState } from "../store/rootReducer";

interface AllTheProvidersProps {
  children?: any;
}

const AllTheProviders = ({ children }: AllTheProvidersProps) => {
  const middlewares: any[] = [];
  const mockStore = configureStore(middlewares);

  const intialState: RootState = {
    preferences: {
      isPreferencesClicked: false,
      theme: ThemeEnum.NIGHT,
      isOneColumnLayout: true,
    },
    stories: {
      list: [],
      startAt: 0,
      pending: false,
      error: null,
    },
    user: {
      info: null,
      pending: false,
      error: null,

      stories: {
        list: [],
        startAt: 0,
        pending: false,
        error: null,
      },
    },
  };
  const store = mockStore(intialState);

  return (
    <Provider store={store}>
      <Router>
        <ThemeProvider>{children}</ThemeProvider>
      </Router>
    </Provider>
  );
};

const customRender = (ui: any, options?: any) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
