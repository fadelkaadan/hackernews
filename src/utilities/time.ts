export const covnvertUnixTimeToDate = (unixTime: number): Date => {
  const milliseconds = unixTime * 1000;
  return new Date(milliseconds);
};

export const months: any = {
  "0": "January",
  "1": "February",
  "2": "March",
  "3": "April",
  "4": "May",
  "5": "June",
  "6": "July",
  "7": "August",
  "8": "September",
  "9": "October",
  "10": "November",
  "11": "December",
};
