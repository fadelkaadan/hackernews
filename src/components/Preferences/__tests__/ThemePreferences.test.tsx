import React from "react";
import { render, screen, RenderResult } from "../../../utilities/test-utils";

import ThemePreferences from "../ThemePreferences";

describe("ThemePreferences", () => {
  it("Should match the snapshot", () => {
    const { container }: RenderResult = render(<ThemePreferences />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Should render 3 preference buttons", () => {
    render(<ThemePreferences />);
    const buttons = screen.getAllByRole("button");
    expect(buttons.length).toBe(2);
  });
});
