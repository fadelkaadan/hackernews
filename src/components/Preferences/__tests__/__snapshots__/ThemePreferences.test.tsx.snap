// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemePreferences Should match the snapshot 1`] = `
<div
  class="sc-bdnylx cjjxQC"
>
  <button
    class="sc-gtssRu sc-dlnjPT hOqmoc fKGzSm"
  />
  <button
    class="sc-gtssRu sc-hKFyIo hOqmoc ggcmdP"
  />
  <button
    class="sc-gtssRu sc-eCApGN hOqmoc iIdOnQ"
  />
</div>
`;
